irispy/iris_tools.py

```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
     return iris_response
 
 
+def fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs):
+    """
+    To calculate the coefficients of best-fit time function for throughput,
+    for which there are two modes:
+    1. Perform fit: supply xput and single element ``cal_coeffs``.
+    2. Apply fit: supply full ``cal_coeffs``.
+
+    The procedure involved in this function is as follows:
+    1. The time difference (in years) is computed from the time_obs and time_cal_coeffs.
+    2. A least-sqaures fit is performed to determine the best fit for the time-dependent
+    effective areas given the time difference.
+
+    Parameters
+    ----------
+    time_obs: a numpy array of floats
+    - Times of datapoints.
+
+    time_cal_coeffs: a numpy array of floats (with exactly two columns)
+    - Start and end times of intervals of constant ``cal_coeffs[i]``.
+
+    cal_coeffs: a numpy array of floats (with at least two columns)
+    - Coefficients of best-fit function.
+
+    Returns
+    -------
+    Yields the effective areas as values of fit for times ``time_obs``.
+
+    """
+    # Convert the observation time into an astropy ``time.time`` object
+    time_obs = np.array([parse_time(time_obs, format = 'utime') for t in time_obs])
+
+    size_time_cal_coeffs = time_cal_coeffs.shape
+    size_cal_coeffs = cal_coeffs.shape
+
+    if size_time_cal_coeffs[1] != 2 or size_cal_coeffs[1] < 2:
+        # Raise ValueError as time coefficient have the wrong format.
+        raise ValueError("Incorrect number of elements either in time_cal_coeffs or in cal_coeffs.")
+
+    # Some time transformations.
+    # Convert the time_cal_coeffs given in the .geny file into a ``datetime.datetime`` object
+    # called t_cal_coeffs, so that the time differences will be in days...
+    t_cal_coeffs_flat = time_cal_coeffs.flatten()
+    t_cal_coeffs = [parse_time(x, format = 'utime') for x in t_cal_coeffs_flat]
+    t_cal_coeffs = np.array(t_cal_coeffs).reshape(size_time_cal_coeffs)
+
+    # For loop for carrying out the least-squares fit and computation of fit output.
+    count = 0
+    for i in time_obs:
+        aux_cal_coeffs = np.zeros(2*size_time_cal_coeffs[0])
+
+        # Exponent for transition between exp.decay intervals.
+        transition_exp = 1.5
+
+        fit_out = np.zeros(len(list(time_obs)), dtype=np.float64)
+
+        # Looking for the closest time in the calibration time intervals.
+        # Differences are given in years before passing to the next stage.
+        t_diff = time_obs - t_cal_coeffs
```

Comment:
Consistent with required change raised in above comment, this line would have to become
```python
t_diff = t - t_cal_coeffs
```